polus-ftl-label-plugin/src/ftl.pyx

```diff
@@ -0,0 +1,966 @@
+    # The number of edges discovered
+    cdef unsigned long count
+
+    ''' Stride information '''
+    cdef unsigned int [:] strides = np.ones(ndim,dtype=np.uint32)
+    for i in range(1,ndim):
+        for j in range(ndim-i):
+            strides[j] = strides[j]*np.uint32(shape[ndim - i])
+    cdef Py_ssize_t last_stride = shape[ndim-1]
+    positions = strides[0]*shape[0]//last_stride
+        
+    '''
+    
+    Loop through all points and find the start and stopping edges of
+    consecutive nonzero values. The way this works is that pixels are laid
+    out linearly in memory according to the last dimension of the matrix.
+    So, in a 2-d matrix pixels in the same row are immediately next to
+    each other in memory. In a 3-d matrix, pixels in the z-dimension are
+    next to each other in memory. Since the input is required to be 8-bit
+    data (boolean or uint8), then mapping data directly to a vector allows
+    8x more data to be processed per clock cycle. Otherwise each clock
+    cycle is used to process one 8-bit piece of data.
+
+    The outtermost loop is a linear index of the upper dimensions. So, if
+    the matrix is 3-d with dimensions (128x128x64), then there are 128x128
+    positions to evaluate. Each position is starting point for a new line
+    of pixels along the last dimension.
+    
+    '''
+    for p in range(positions):
+
+        # Calculate the upper dimension values
+        dim = position
+        for i in range(ndim-1):
+            dim = dim//strides[i]
+            coords[i] = dim
+            dim = position - dim*strides[i]
+
+        # The following code ignores the first pixel, so check to see if
+        # the first pixel has a value.
+        if image[position] > 0:
+            on_obj=True
+            for i in range(ndim-1):
+                output.push_back(coords[i])
+            output.push_back(0)
+        
+        '''
+        
+        The following loop is designed for speed. It analyzes 15 pixels at
+        a time, and if all 15 pixels have the same value then it quickly
+        escapes to the next iteration of the loop. It stops loops when it
+        gets less than 15 pixels from the end of the line of pixels so that
+        it doesn't run into the next line of pixels.
+        
+        '''
+        r = 0 # manualy register the pixel index, can be optimized with modulo outside the loop
+        # print('Starting fast loop position: {}'.format(position))
+        for n in range(0,last_stride-15,15):
+            # Load 16 pixels and compare against the neighbors (technically only checks 15 pixels)
+            v      = _mm_loadu_si128(<__m128i*>&image[position])   # Load 16 pixels.
+            mask   = _mm_cmpgt_epi8(v,_mm_setzero_si128())         # Generate 16x8 bit mask.
+            mask   = _mm_xor_si128(mask,_mm_slli_si128(mask,1))    # Shift bits to find leading and trailing edges
+
+            # advance the memory position
+            position += 15
```

Comment:
Is 15 related to the register size of 16x8bits? If not, playing with this empirical value might impact  the performance. It is good to calibrate this value for your code.

Reply:
15 is actually related to how I search for edges of pixel runs. Basically, I take a set of 16 pixels in two vectors, shift one of the vectors by one value, then use xor to determine where the pixels are different from each other. This tells me where the edges are in roughly two clock cycles over 15 pixels.

For example, if I have a set of binary values:
00111111000011100

I can shift all values by one unit and compare with xor:
00111111000011100 - original
00011111100001110 - shifted
00100000100010010 - xor

The reason why 15 is the magic number is because one of the pixels gets shifted out of the vector, so I can't use that. The position of the non-zero values are then the run length encoded start and stop positions. In this example, I can store the first run of 1's as [2,8] rather than keeping all of the individual pixels. The reason why this is very fast using intrinsics is because I can run these operations 15 pixels at a time and instantly identify the leading and trailing edges. Even better, when all pixels are 0 or 1, I can just go to the next iteration of the loop.